### ISSUE TITLE
Question about "FairseqEncoderDecoder output dimension" & Decoder.get_normalized_probs function

### ISSUE THREAD

**Issue by steventan0110** [pytorch/fairseq#2591]
Why FairseqDecoder get_normalized_probs uses output[0] as input for softmax function? (it results in dimension mismatch for my loss calculation)

I was trying to implement my own model (LSTM based, with encoder side attention). In the documentation, I learned that `forward` function of FairseqEncoderDecoderModel should return an output of shape (batch_size, target_length, vocab_size), which is very intuitive. However, as I made my model to return this shape, I constantly get dimension mismatched issue. Then I checked the source code for `criterion` and found that it called `decoder.get_normalized_output` function. This function use net_output[0] as input for a softmax function, which means it only uses the first batch of my output, and that result in a mismatch for later loss calculation. Is there something wrong about my understanding of FairseqEncoderDecoderModel's output? Thank you for your help in advance.

#### Code
the compute_loss function called by cross_entropy.py in fairseq:
```
loss, _ = self.compute_loss(model, net_output, sample, reduce=reduce)

def compute_loss(self, model, net_output, sample, reduce=True):
        lprobs = model.get_normalized_probs(net_output, log_probs=True)
        #print('size before view:', lprobs.size())
        lprobs = lprobs.view(-1, lprobs.size(-1))
        #print('size after view:', lprobs.size())
        target = model.get_targets(sample, net_output).view(-1)
        #print('target dimension:', target.size())
        loss = F.nll_loss(
            lprobs,
            target,
            ignore_index=self.padding_idx,
            reduction='sum' if reduce else 'none',
        )
        return loss, loss
```
Then notice the `get_normalized_probs` function, which eventually called following lines:
```
def get_normalized_probs(
        self,
        net_output: Tuple[Tensor, Optional[Dict[str, List[Optional[Tensor]]]]],
        log_probs: bool,
        sample: Optional[Dict[str, Tensor]] = None,
    ):
        """Get normalized probabilities (or log probs) from a net's output."""

        logits = net_output[0]

        if log_probs:
            return utils.log_softmax(logits, dim=-1, onnx_trace=self.onnx_trace)
        else:
            return utils.softmax(logits, dim=-1, onnx_trace=self.onnx_trace)
```
if output has size (batch_size, target_length, vocab_size), then `logits` above will have size (target_length, vocab) 
But in the F.nll_loss function above, the target variable have dimension (batch_size*target_length) and by pytorch official documents, the lprobs should have dimension (batch_size*target_length, vocab), which is not what I have here (target_length, vocab).

#### What have you tried?

I tried to change my output dimension to (1, batch_size*target_length, vocab) and the code works, but I'm confused because this is not intuitive for an output dimension. And I still don't understand why the `logits` refer to the output[0].

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0) 1.6
 - OS (e.g., Linux): linux
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): 
```
fairseq-train /home/steven/Documents/GITHUB/DQN/data/multilingual/glgpor_eng \
--arch attention_lstm  --encoder-dropout 0.2 --decoder-dropout 0.2   --optimizer adam \
--lr 0.005 --lr-shrink 0.5 --max-tokens 12000   \
--user-dir /home/steven/Documents/GITHUB/DQN/src \
--save-dir /home/steven/Documents/GITHUB/DQN/checkpoints/multilingual/glgpor_eng
```
 - Python version: 3.7
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: RTX2070, default config
 - Any other relevant information:

**Comment by steventan0110** [comment]
Thank you! I pass in None as the second element (optional output) of the tuple and solved the problem.
